packages/@aws-cdk/core/lib/arn.ts

```diff
@@ -268,3 +284,45 @@ function parseToken(arnToken: string, sep: string = '/', hasName: boolean = true
     return { partition, service, region, account, resource, resourceName, sep };
   }
 }
+
+
+/**
+ * Validate that a string is either unparseable or looks mostly like an ARN
+ */
+function parseArnShape(arn: string): 'unparseable' | string[] {
+  const components = arn.split(':') as Array<string>;
+  const looksLikeArn = arn.startsWith('arn:') && components.length >= 6 && components.length <= 7;
+  if (Token.isUnresolved(arn) && !looksLikeArn) {
+    return 'unparseable';
+  }
+  // If the ARN merely contains Tokens, but otherwise *looks* mostly like an ARN,
+  // it's a string of the form 'arn:${partition}:service:${region}:${account}:abc/xyz'.
+  // Parse fields out to the best of our ability.
+  // Tokens won't contain ":", so this won't break them.
+
+  if (components.length < 6) {
+    throw new Error('ARNs must have at least 6 components: ' + arn);
+  }
+
+  // If the ARN merely contains Tokens, but otherwise *looks* mostly like an ARN,
+  // it's a string of the form 'arn:${partition}:service:${region}:${account}:abc/xyz'.
+  // Parse fields out to the best of our ability.
+  // Tokens won't contain ":", so this won't break them.
+  const [arnPrefix, , service, , , sixth] = components;
+
+  if (arnPrefix !== 'arn') {
+    throw new Error('ARNs must start with "arn:": ' + arn);
+  }
```

Comment:
unreachable because of line 295. 
Maybe remove the `startsWith()` conditional in the looksLikeArn and let the error on this line show up.

Reply:
It looks like it is, but it's not.

| Unresolved | LooksLikeArn | Behavior |
|------|-------|-----------|
| false | false | Will trip `arnPrefix` check |
| false | true | Will not trip `arnPrefix` check |
| true | false | Will be treated as unparseable |
| true | true | Check is superfluous |

I will change the order somewhat to make it more clear though.